src/main/kotlin/com/coder/gateway/sdk/CoderRestClientService.kt

```diff
@@ -218,3 +250,203 @@ class CoderRestClient(var url: URL, var token: String,
         }
     }
 }
+
+fun coderSocketFactory(settings: CoderSettingsState) : SSLSocketFactory {
+    if (settings.tlsCertPath.isBlank() || settings.tlsKeyPath.isBlank()) {
+        return SSLSocketFactory.getDefault() as SSLSocketFactory
+    }
+
+    val certificateFactory = CertificateFactory.getInstance("X.509")
+    val certInputStream = FileInputStream(expandPath(settings.tlsCertPath))
+    val certChain = certificateFactory.generateCertificates(certInputStream)
+    certInputStream.close()
+
+    // ideally we would use something like PemReader from BouncyCastle, but
+    // BC is used by the IDE.  This makes using BC very impractical since
+    // type casting will mismatch due to the different class loaders.
+    val privateKeyPem = File(expandPath(settings.tlsKeyPath)).readText()
+    val start: Int = privateKeyPem.indexOf("-----BEGIN PRIVATE KEY-----")
+    val end: Int = privateKeyPem.indexOf("-----END PRIVATE KEY-----", start)
+    val pemBytes: ByteArray = Base64.getDecoder().decode(
+        privateKeyPem.substring(start + "-----BEGIN PRIVATE KEY-----".length, end)
+            .replace("\\s+".toRegex(), "")
+    )
+
+    var privateKey : PrivateKey
+    try {
+        val kf = KeyFactory.getInstance("RSA")
+        val keySpec = PKCS8EncodedKeySpec(pemBytes)
+        privateKey = kf.generatePrivate(keySpec)
+    } catch (e: InvalidKeySpecException) {
+        val kf = KeyFactory.getInstance("EC")
+        val keySpec = PKCS8EncodedKeySpec(pemBytes)
+        privateKey = kf.generatePrivate(keySpec)
+    }
+
+    val keyStore = KeyStore.getInstance(KeyStore.getDefaultType())
+    keyStore.load(null)
+    certChain.withIndex().forEach {
+        keyStore.setCertificateEntry("cert${it.index}", it.value as X509Certificate)
+    }
+    keyStore.setKeyEntry("key", privateKey, null, certChain.toTypedArray())
+
+    val keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm())
+    keyManagerFactory.init(keyStore, null)
+
+    val sslContext = SSLContext.getInstance("TLS")
+
+    val trustManagers = coderTrustManagers(settings.tlsCAPath)
```

Comment:
Is it a problem that we only use `tlsCAPath` here if we did not return early above when `tlsCertPath` or `tlsKeyPath` are blank?  Thinking of a use case were someone only sets `tlsCAPath` but not the others.  I see we separately call `coderTrustManagers()` in the rest client so maybe it works there, but possibly the binary download could have issues since we are only using the socket factory there.